internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+				RemainingItems: output.GetRemainingItems(),
+			}
+
+			this.run(backgroundOutput, false)
+
+			if backgroundOutput.Error != nil {
+				log.Error("Failure while running asynchronous job: '%v'.", output.Error)
+			}
+		}()
+
+		close(done)
+		return &output
+	} else {
+		go func() {
+			this.run(&output, true)
+			if output.Error != nil {
+				log.Error("Failure while running asynchronous job: '%v'.", output.Error)
+			}
+
+			close(done)
+		}()
+	}
+
+	return &output
+}
+
+// Given a partial job output, run job processes the remaining items in a thread safe manner.
+// When not updating the output, only the run time and errors will be available.
+func (this *Job[InputType, OutputType]) run(output *JobOutput[InputType, OutputType], updateOutput bool) {
+	if len(output.GetRemainingItems()) == 0 {
+		return
+	}
+
+	noLockWait := true
+	if this.LockKey != "" {
+		noLockWait = lockmanager.Lock(this.LockKey)
+		defer lockmanager.Unlock(this.LockKey)
+	}
+
+	// The context has been canceled while waiting for a lock, abandon this job.
+	if this.Context.Err() != nil {
+		return
+	}
+
+	var err error = nil
+
+	// If we had to wait for the lock, then check again for stored records.
+	if !noLockWait && !this.OverwriteRecords && this.RetrieveFunc != nil {
+		partialResults := []OutputType{}
+		remainingItems := []InputType{}
+		partialResults, remainingItems, err = this.RetrieveFunc(output.GetRemainingItems())
+		if err != nil {
+			output.SetError(fmt.Errorf("Failed to re-check record storage before run: '%w'.", err))
+			return
+		}
+
+		output.SetRemainingItems(remainingItems)
+
+		if updateOutput {
+			// Collect the partial records from storage.
+			output.SetResultItems(append(output.GetResultItems(), partialResults...))
+		}
+	}
+
+	if len(output.GetRemainingItems()) == 0 {
+		return
+	}
+
+	type PoolResult struct {
+		Input   InputType
+		Result  OutputType
+		RunTime int64
+		Error   error
+	}
+
+	poolResults, _, err := util.RunParallelPoolMap(this.PoolSize, output.GetRemainingItems(), this.Context, func(workItem InputType) (PoolResult, error) {
+		workItemKey := ""
+		if this.WorkItemKeyFunc != nil {
+			workItemKey = this.WorkItemKeyFunc(workItem)
+		}
+
+		// Optionally lock this id so we don't work on an item multiple times.
+		if workItemKey != "" {
+			lockmanager.Lock(workItemKey)
+			defer lockmanager.Unlock(workItemKey)
+		}
+
+		if this.Context.Err() != nil {
+			return PoolResult{}, nil
+		}
+
+		// Check storage for a record.
+		if !this.OverwriteRecords && this.SingleRetreiveFunc != nil {
+			result, found, err := this.SingleRetreiveFunc(workItem)
+			if err != nil {
+				return PoolResult{
+					Input: workItem,
+					Error: fmt.Errorf("Failed to retrieve an individual record from storage: '%w'.", err),
+				}, nil
+			}
+
+			if found {
+				return PoolResult{workItem, result, 0, nil}, nil
+			}
+		}
+
+		// Nothing cached, compute the job.
+		startTime := timestamp.Now()
+
+		var errs error = err
```

Comment:
Just a few lines above, you have an error that you didn't join.
I don't see why you need to start joining errors here.